src/XMakeBuildEngine/BackEnd/Components/Communications/NodeProviderOutOfProc.cs

```diff
@@ -69,39 +69,33 @@ public int AvailableNodes
 
         /// <summary>
         /// Magic number sent by the host to the client during the handshake.
-        /// Derived from the binary timestamp to avoid mixing binary versions.
+        /// Derived from the binary timestamp to avoid mixing binary versions,
+        /// Is64BitProcess to avoid mixing bitness, and enableNodeReuse to
+        /// ensure that a /nr:false build doesn't reuse clients left over from
+        /// a prior /nr:true build.
         /// </summary>
-        internal static long HostHandshake
+        /// <param name="enableNodeReuse">Is reuse of build nodes allowed?</param>
+        internal static long GetHostHandshake(bool enableNodeReuse)
         {
-            get
-            {
-                long baseHandshake = Constants.AssemblyTimestamp;
-                if (Environment.Is64BitProcess)
-                {
-                    unchecked
-                    {
-                        baseHandshake = baseHandshake ^ (long)0x8b8b8b8b8b8b8b8b;
-                    }
-                }
-
-                baseHandshake = CommunicationsUtilities.GenerateHostHandshakeFromBase(baseHandshake, ClientHandshake);
-                return baseHandshake;
-            }
+            long baseHandshake = Constants.AssemblyTimestamp;
+
+            baseHandshake = baseHandshake*17 + Environment.Is64BitProcess.GetHashCode();
+
+            baseHandshake = baseHandshake*17 + enableNodeReuse.GetHashCode();
+
+            return CommunicationsUtilities.GenerateHostHandshakeFromBase(baseHandshake, GetClientHandshake());
         }
 
         /// <summary>
         /// Magic number sent by the client to the host during the handshake.
         /// Munged version of the host handshake.
         /// </summary>
-        internal static long ClientHandshake
+        internal static long GetClientHandshake()
         {
-            get
-            {
-                // Mask out the first byte. That's because old
-                // builds used a single, non zero initial byte,
-                // and we don't want to risk communicating with them
-                return (Constants.AssemblyTimestamp ^ Int64.MaxValue) & 0x00FFFFFFFFFFFFFF;
-            }
+            // Mask out the first byte. That's because old
+            // builds used a single, non zero initial byte,
+            // and we don't want to risk communicating with them
+            return (Constants.AssemblyTimestamp ^ long.MaxValue) & 0x00FFFFFFFFFFFFFF;
```

Comment:
Int64 might actually be more self-documenting here, since it mentions the mask size.Not worth changing the commit though.